File: 2021/02/202102.exs
# declare the atoms (for safety)
_cmds = %{forward: 1, down: 2, up: 3}

parse_cmd = fn s ->
  case String.split(s, " ") do
    [x, y] -> {String.to_existing_atom(x), String.to_integer(y)}
    _ -> {}
  end
end

parse = fn x ->
  String.trim(x)
  |> String.split("\n")
  |> Enum.map(parse_cmd)
end

{:ok, input} = File.read("input.txt")
data = parse.(input)

move = fn
  {:forward, num}, {p, d} -> {p + num, d}
  {:down, num}, {p, d} -> {p, d + num}
  {:up, num}, {p, d} -> {p, d - num}
end

{p, d} = Enum.reduce(data, {0, 0}, move)

IO.puts("PART1: #{p * d}")

move2 = fn
  {:forward, num}, {p, d, a} -> {p + num, d + a * num, a}
  {:down, num}, {p, d, a} -> {p, d, a + num}
  {:up, num}, {p, d, a} -> {p, d, a - num}
end

{p, d, _a} = Enum.reduce(data, {0, 0, 0}, move2)

IO.puts("PART2: #{p * d}")

File: 2021/02/202102.py
tds = """forward 5
down 5
forward 8
up 3
down 8
forward 2"""


def parse(cmds):
    return [(c, int(v)) for c, v in
            [s.split() for s in cmds.strip().splitlines()]]


def part1(cmds):
    hp, d = 0, 0
    for c, v in cmds:
        if c == 'forward':
            hp += v
        elif c == 'down':
            d += v
        elif c == 'up':
            d -= v
    return hp * d


td = parse(tds)
data = parse(open("input.txt").read())
assert part1(td) == 150
print("PART1:", part1(data))


def part2(cmds):
    hp, d, aim = 0, 0, 0
    for c, v in cmds:
        if c == 'forward':
            hp += v
            d += aim * v
        elif c == 'down':
            aim += v
        elif c == 'up':
            aim -= v
    return hp * d


assert part2(td) == 900
print("PART2:", part2(data))
